packages/compass-editor/src/codemirror/document-autocompleter.ts

```diff
@@ -0,0 +1,82 @@
+import type {
+  CompletionContext,
+  CompletionSource,
+} from '@codemirror/autocomplete';
+import { syntaxTree } from '@codemirror/language';
+import { completeAnyWord, ifIn } from '@codemirror/autocomplete';
+import { completer, wrapField } from '../autocompleter';
+import { languageName } from '../json-editor';
+
+const completeWordsInString = ifIn(['String'], completeAnyWord);
+
+function resolveTokenAtCursor(context: CompletionContext) {
+  return syntaxTree(context.state).resolveInner(context.pos, -1);
+}
+
+type Token = ReturnType<typeof resolveTokenAtCursor>;
+
+function isJSONPropertyName(token: Token): boolean {
+  return (
+    // Cursor is currently on the valid property name in the object
+    token.name === 'PropertyName' ||
+    // Cursor is possibly on the invalid property name as indicated by the
+    // previous sibling being a property or an open bracket and not a
+    // `PropertyName`, which would be the case for property value
+    (token.type.isError &&
+      ['Property', '{'].includes(token.prevSibling?.name ?? ''))
+  );
+}
+function isJavaScriptPropertyName(token: Token): boolean {
+  return (
+    // Cursor is currently inside a property
+    token.parent?.name === 'Property' &&
+    // There is no previous sibling or it's an opening bracket (indicating
+    // computed property)
+    (!token.prevSibling || token.prevSibling.name === '[')
+  );
+}
+
+/**
+ * Autocompleter for the document object, only autocompletes field names in the
+ * appropriate format (either escaped or not) both for javascript and json modes
+ */
+export const createDocumentAutocompleter = (
+  fields: string[]
+): CompletionSource => {
+  const completions = completer('', { fields, meta: ['field:identifier'] });
+
+  return (context) => {
+    const token = resolveTokenAtCursor(context);
+
+    const shouldEscapeProperty =
```

Comment:
Very minor thing, maybe call this `shouldAlwaysEscapeProperty` or similar, since `wrapField()` can escape the string even if this is `false`?

Reply:
Yes, good suggestion, thank you!